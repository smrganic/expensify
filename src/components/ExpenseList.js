import React from "react"
import ExpenseItem from "./ExpenseListItem"
import { connect } from "react-redux"
import selectExpenses from "../redux/selectors/expenses"

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                    props.expenses
                        .map(
                            (expense) =>
                                (<ExpenseItem key={expense.id} {...expense} />)
                        )
                )
        }
    </div>
)

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

export default ConnectedExpenseList