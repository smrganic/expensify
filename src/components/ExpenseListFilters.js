import React from "react"
import { connect } from "react-redux"
import { DateRangePicker } from "react-dates"
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../redux/actions/filters"
import moment from "moment"

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (event) => {
        this.props.setTextFilter(event.target.value)
    }

    onSortChange = (event) => {
        if (event.target.value === "date") {
            this.props.sortByDate()
        } else {
            this.props.sortByAmount()
        }
    }

    render() {
        return (
            <div>
                <input
                    txpe="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId={moment().toString()}
                    endDate={this.props.filters.endDate}
                    endDateId={moment().toString()}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({ filters: state.filters })

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)