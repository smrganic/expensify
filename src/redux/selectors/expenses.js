import moment from "moment"

// Sorting based on dates & text
// Default sort uses sort function where -1 return represents a and 1 respresents b
const getVisibleElements = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
        return textMatch && startDateMatch && endDateMatch
    }).sort((a, b) => {
        if (sortBy === "date")
            return a.createdAt > b.createdAt ? -1 : 1
        else if (sortBy === "amount")
            return a.amount > b.amount ? -1 : 1
    })
}

export default getVisibleElements