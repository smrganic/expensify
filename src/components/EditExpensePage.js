import React from "react"
import { connect } from "react-redux"
import ExpenseForm from "./ExpenseForm"
import { editExpense, removeExpense } from "../redux/actions/expenses"

export class EditExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push("/")
    }

    onClick = (event) => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push("/")
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick}>Remove</button>
            </div>
        )
    }
}

// Implicitly returns an object
const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense({ id }))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)