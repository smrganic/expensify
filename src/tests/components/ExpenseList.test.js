import React from "react"
import { shallow } from "enzyme"
import { ExpenseList } from "../../components/ExpenseList"
import expenses from "../mock-data/expenses"

test("Should render ExpenseList with expenses.", () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})

test("Should render ExpenseList with no expenses.", () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />)
    expect(wrapper).toMatchSnapshot()
})