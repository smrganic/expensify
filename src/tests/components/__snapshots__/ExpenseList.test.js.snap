// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with expenses. 1`] = `
<div>
  <ExportListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExportListItem
    amount={1000000}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExportListItem
    amount={4500}
    createdAt={345600000}
    description="Credit card"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`Should render ExpenseList with no expenses. 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
